### DISCUSSION TITLE
Potentially ceasing development of Paste

### DISCUSSION THREAD

**Discussion by cdent** [cdent/paste#91] [Announcements]
paste uses a lot of deprecated functionality, much of it related to old libraries like `cgi` and `cgitb`.

When Python 3.13 becomes the main Python release this deprecated functionality will be removed and without a fair bit of work paste will stop working.

I personally do not think we should continue to maintain paste. It is an old tool using old technology that is no longer aligned with modern techniques or tools.

I would like us to consider winding it down but I'm not certain who should be involved in the discussion so pinging some of the people who have made contributions to the project over the last while: @a-detiste , @amitmarkel , @benjaminp , @Cito , @cjwatson , @CyrilRoelandteNovance , @blueyed , @brondsem , @hugovk , and of course @ianb . 

Please register your opinion.

If you feel like paste should carry on living, and want to volunteer to take over maintenance I'm very willing to transfer ownership.

I also maintain pastescript and feel it should end too, so if someone wants to take them as a package deal that would be great.

**Comment by hugovk** [comment]
I personally don't need to use Paste, but if people still need it, one option is to use the community fork of cgi/cgitb:

* https://pypi.org/project/legacy-cgi/ 
* https://github.com/jackrosenthal/python-cgi

**Reply by cdent** [reply]
Thanks, I'll keep those in mind depending on what direction this conversation goes.

**Comment by a-detiste** [comment]
We also have zombie-imp in Debian. We could add this one to our collection
of "dead batteries".

Le ven. 5 avr. 2024, 20:02, Hugo van Kemenade ***@***.***> a
écrit :

**Comment by Cito** [comment]
Personally I don't use Paste and agree that this looks like the right moment to end support for paste.

However, keep in mind that paste was extremely popular and several important projects still depend on paste. For instance,

- OpenStack [Glance](https://github.com/openstack/glance), [Neutron](https://github.com/openstack/neutron), [Nova](https://github.com/openstack/nova) etc.
- [galaxy](https://github.com/galaxyproject/galaxy)
- ~~[Pyramid](https://github.com/Pylons/pyramid)~~ (edit: current version doesn't depend on paste any more)
- [PySAML2](https://github.com/IdentityPython/pysaml2)
- [Zope](https://github.com/zopefoundation/Zope)

Maybe notify the most important projects and ask if they can move away from paste or anybody wants to continue to maintain paste.

I had a similar problem with [Webware for Python](https://webwareforpython.github.io/w4py3/) and solved it by incorporating `cgi.FieldStorage` into Webware.

**Reply by a-detiste** [reply]
From telemetry alone this package is thriving: https://qa.debian.org/popcon.php?package=paste

**Reply by cdent** [reply]
Yeah, the reason I got involved with maintaining paste at all was because of its continued use in OpenStack with which I was _heavily_ involved for a few years.

While the cgi and cgitb issues provide a good moment to say "let's spin this down" in no small part it is a convenient excuse for something I think should have happened a long time ago. Paste is definitely no longer a good choice and hasn't been for a long time. I never really understood its initial popularity and definitely don't understand its continued popularity.

But yeah, it's not right to kill it off without first consulting some of its major consumers. Part of the reason for starting this thread was to gather more info on who those consumers might be and see about bringing them into the conversation. I'll seek out some people to invite in. If any of the rest of you have people you can contact please do.

**Reply by cjwatson** [reply]
I'm just going by https://www.wheelodex.org/projects/paste/rdepends/ etc., but I think the only bits of Zope that use Paste and friends any more are themselves pretty well deprecated, e.g. `zope.server`.

My own interest in Paste ended when I stopped working for Launchpad (which doesn't use it directly, but it has a few dependencies that do).  Maybe @javacruft would be interested on behalf of OpenStack, or @jelmer since Loggerhead uses Paste.

**Reply by jelmer** [reply]
Thanks for adding me to the thread. I'm interested in keeping Loggerhead functional, but am guessing that migrating away from paste is probably going to be the better option in the long term.

**Reply by javacruft** [reply]
Thanks @cjwatson  - @fungi already raised this for discussion on the OpenStack ML.

**Comment by brondsem** [comment]
There's also https://github.com/Pylons/pastedeploy/ which is still maintained.  I think it depends on Paste right?  Mentioning recent committers to that repo: @mmerickel @fmigneault @digitalresistor

**Reply by fmigneault** [reply]
It is installed as a namespace package inside `paste`. I'm not familiar enough with the internals to know "how much" it depends on `paste`. 

I've been mostly using `pserve` with `pyramid`.
I have at least 2 projects running with `egg:Paste#urlmap`, `egg:Paste#static` and `egg:PasteDeploy#prefix`. 
Not sure if those can be easily replicated without `paste`?

**Reply by mmerickel** [reply]
> I have at least 2 projects running with egg:Paste#urlmap, egg:Paste#static and egg:PasteDeploy#prefix.

Just fyi the `egg:Paste#urlmap` has a decoupled version called [rutter](https://pypi.org/project/rutter/) that I help maintain that is a drop in replacement. And obviously pastedeploy works without paste. I do not know off the top of my head if there is a replacement for `egg:Paste#static`.

`pserve` and pyramid and pastedeploy have no dependencies on paste themselves.

**Comment by cdent** [comment]
So it's been about 2 weeks and the summary so far seems to be:

1. There are a few projects which definitely depend on paste (e.g., some OpenStack projects)
2. There are a few projects which have associations with paste, either by namespace or use of a small amount of functionality but themselves might be considered in a state of "time for an upgrade"
3. https://pypi.org/project/legacy-cgi/ could get Paste paste the immediate 3.13 concern but further investigation is required to know if there are other areas that might be a problem.
4. Thus far nobody has stepped up to say "yeah, I/we will take over maintenance".

That sound about right?

**Comment by mmerickel** [comment]
Hey folks -- with respect to pyramid and pastedeploy -- these projects aren't going anywhere but they also do not depend on paste anymore. So from that perspective I don't have a strong opinion on the future of the paste project itself.

WIth respect to the removal of the cgi module, webob also is affected by this this issue - the plan for it is to vendor the relevant parts of cgi into itself.

The dependency on the paste namespace is pretty meaningless, as pastedeploy works fine without paste installed.

**Comment by Cito** [comment]
I think the only needed parts are the cgi.FieldStorate and cgitb.Hook classes, which could be easily included in paste. Would be good to have one *last* release with this change, and then declare the project as officially retired and unmaintained. The Galaxy and OpenStack folks should be also notified, but they will then have sufficient time to adapt.

**Reply by cjwatson** [reply]
Most uses of `cgi.FieldStorage` should probably be handled by switching to [multipart](https://pypi.org/project/multipart/), which is less buggy anyway, rather than by trying to keep that code alive in `paste`.  [PEP 594](https://peps.python.org/pep-0594/#cgi) recommends this too.

**Reply by Cito** [reply]
If you want to keep paste alive, maybe that's a good idea. But if you want to retire paste, it's better to release one last, fully compatible version - and compatible means supporting FieldStorage.

**Comment by Cito** [comment]
I have now added PR #94 and #95 which include the relevant parts of cgi and cgitb and #93 which handles a few more deprecation errors.

**Reply by cdent** [reply]
Thanks for those, I'll check them out properly in the next couple of days.

I agree with the strategy of "release one last, fully compatible version", especially given that no one has stepped forward to take over maintenance.

**Comment by brondsem** [comment]
I might be willing to maintain Paste & PasteScript.  Apache Allura and an internal repo I maintain use `egg:Paste#http` and PasteScript

I think there are probably many projects using Paste and PasteScript beyond the few identified already, so I tried to get a feel for that.  https://www.wheelodex.org/projects/paste/rdepends/ lists other libraries, but many are very old.  https://github.com/cdent/paste/network/dependents seems to be extremely inaccurate.  https://github.com/search?q=%22paste%3D%3D3.6%22+OR+%22paste%3D%3D3.7%22+OR+%22paste%3D%3D3.8%22+OR+%22paste%3D%3D3.9%22&type=code shows repos that rely on the past few versions of Paste (if they have it explicitly pinned).  There's likely non-public projects that use Paste too.  Similar for PasteScript

It sounds like the next release will support Python 3.13, and perhaps it'll "just work" on 3.14 and higher too.  So it may be many years before any users of Paste/PasteScript even encounter issues with it being unsupported.  Maybe they will be interested in helping if/when they run into issues in the future.

Are there any recommendations we can make for alternatives to Paste and PasteScript?  I don't know offhand of a direct replacement., especially since it does so many things.  And paste script commands are likely used in project's docs, Makefiles, Docker images etc, which isn't as easy to migrate.  

Pulling those thoughts together: for myself and for other projects, the easiest thing right now is to just keep using Paste/PasteScript as-is.  And in the future it may be easier to keep doing small maintenance on Paste/PasteScript than to migrate a big project to something different.  But that is a while away, so hard to know.

@cdent if you think you won't be active going forward, maybe consider adding me as an owner now so there's a better chance of continuity (avoiding risk of a hard fork in the future).  But at the same time, I can't make any promises that I'll necessarily be active in the future when maintenance is needed next 😄

**Reply by cdent** [reply]
> Maybe they will be interested in helping if/when they run into issues in the future.

Part of my thinking when starting this thread was to see if discontinuing support could be a forcing function to either bring people will to support out of the woodwork (some success there) or encourage people to stop using Paste entirely (less successful).

This clearly represents my own biases: I'm not and never have been a paste fan. I took it on as a favour to the OpenStack community while they thought about migration. The maintenance burden has been light, and I rather hoped that the passage of time would move things along. It hasn't.

If you're willing what I think I'll do is deal with the pending pull requests, do a new release and then move the two repos to their own org where I can make you one of the owners and also signal some of the disconnection from me. This may take me some time to get around to, but thanks to the effort from @Cito there's no significant urgency here.

**Reply by hugovk** [reply]
You could consider doing something like Bleach: https://github.com/mozilla/bleach/issues/698

They publicly deprecated the project and made clear the support: updates for security/new Python/egregious bugs, with maybe one release a year.

This prompts other projects to consider their options and plan migration.

**Reply by brondsem** [reply]
Sounds good

**Comment by cdent** [comment]
Okay, here's what's happened:

1. @brondsem is added as an owner here and on pypi of paste and postscript
2. New release and docs for both paste and pastescript have been published
3. Both include mentions of the move and the idea of "deeper" maintenance mode, meaning releases will be even less frequent and PRs only accepted for seriously important stuff unless more active maintainership is found

Please let me know if I've messed anything up :)

**Reply by a-detiste** [reply]
looks good
